sympy/polys/polytools.py

```diff
@@ -6218,8 +6216,44 @@ def sqf_list(f, *gens, **args):
     (2, [(x + 1, 2), (x + 2, 3)])
 
     """
-    return _generic_factor_list(f, gens, args, method='sqf')
-
+    options.allowed_flags(args, ['polys'])
+    opt = options.build_options(gens, args)
+    e = together(f)
+    gen_set = set()
+    coeff = 1
+    facs = []
+    for fac in Mul.make_args(e):
+        b, e = (fac.base, fac.exp) if fac.is_Pow else (fac, S.One)
+        _opt = opt.clone(dict(expand=True))
+        if e < 0:
+            raise PolynomialError("a univariate polynomial expected")
+        try:
+            poly = _poly_from_expr(b, _opt)
+        except PolificationFailed:
+            coeff *= b
+        else:
+            gen_set.add(poly[1].gens)
+            if len(gen_set) > 1:
+                raise PolynomialError("a univariate polynomial expected")
+            cf, sqf_facs = poly[0].sqf_list()
+            coeff *= cf
+            for sqf_fac in sqf_facs:
+                facs.append((sqf_fac[0] , e*sqf_fac[1]))
+    facs = _sorted_factors(facs, 'sqf')
+    if not opt.polys:
+        facs = [(f.as_expr(), k) for f, k in facs]
+    if facs:
+        new_facs = [facs[0]]
+        for i in range(1, len(facs)):
+            l = len(new_facs)
+            if facs[i-1][1] == facs[i][1]:
+                new_facs[l-1] = (Mul(new_facs[l-1][0], facs[i][0]).expand(), new_facs[l-1][1])
```

Comment:
`Mul` and `expand` are intended for expressions. If the factors are `Poly`s, this may not work as expected:
```
In [1]: Mul(Poly(x), Poly(x + 1))
Out[1]: Poly(x, x, domain='ZZ')⋅Poly(x + 1, x, domain='ZZ')
```

Reply:
@jksuom but I have applied `Mul` only after converting into expression. Nevertheless, what should be used instead of `Mul` and `expand` then? Any suggestions?

Reply:
If `opt.polys=True`, there is no conversion. I would change the order so that conversion would come only after the multiplications which can be the standard polynomial multiplications (with `*`). Those are more efficient than computations with expressions.

Reply:
@jksuom I have made the changes you can see.